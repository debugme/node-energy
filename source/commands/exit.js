const command = () => {
  process.exit(0)
}

export { command }
