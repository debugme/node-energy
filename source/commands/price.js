import roundTo from 'round-to'

import { isPositiveInteger } from '../utilities/utilities'
import { DECIMAL_PLACES, DAYS_IN_YEAR, PENCE_IN_POUND, MAX_THRESHOLD } from '../constants/constants'

const getTotalCost = (scratchpad, rateInfo) => {
  const { standing_charge = 0 } = scratchpad
  const { threshold = MAX_THRESHOLD, price } = rateInfo
  const delta = Math.min(scratchpad.balance, threshold)
  scratchpad.balance -= delta
  scratchpad.costInPence += (delta * price)
  scratchpad.costInPence += (DAYS_IN_YEAR * standing_charge)
  return scratchpad
}

const getPlanInfo = (scratchpad, planInfo) => {
  const { usage, vatMultiplier } = scratchpad
  const { supplier, plan, rates, standing_charge } = planInfo
  const options = { costInPence: 0, balance: usage, standing_charge, vatMultiplier }
  const { costInPence } = rates.reduce(getTotalCost, options)
  const costInPounds = costInPence * vatMultiplier / PENCE_IN_POUND
  const totalCost = roundTo(costInPounds, DECIMAL_PLACES)
  const row = { supplier, plan, totalCost }
  scratchpad.table.push(row)
  return scratchpad
}

const parseData = (options) => {
  const { tokens } = options
  if (tokens.length < 2)
    throw new Error('format is "price <annual_usage>"')
  const [, annualUsage] = tokens
  if (!isPositiveInteger(annualUsage))
    throw new Error(`annual usage "${annualUsage}" is not a positive integer amount`)
  return tokens
}

const command = (options) => {
  const { context: { plans, vatMultiplier } } = options
  const [, annualUsage] = parseData(options)
  const usage = parseInt(annualUsage, 10)
  const scratchpad = plans.reduce(getPlanInfo, { usage, vatMultiplier, table: [] })
  scratchpad.table.sort((left, right) => left.totalCost - right.totalCost)
  return scratchpad.table
}

export { command }
