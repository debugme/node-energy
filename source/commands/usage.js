import { isPositiveInteger, findByMatch, findByQuery } from '../utilities/utilities'
import { MONTHS_IN_YEAR, DAYS_IN_YEAR, PENCE_IN_POUND, MAX_THRESHOLD } from '../constants/constants'

const updateUsage = (scratchpad, rate) => {
  const { vatMultiplier } = scratchpad
  const { price, threshold = MAX_THRESHOLD } = rate
  const costPerUnitOfEnergy = (price * vatMultiplier)
  for (let index = 0; index < threshold; ++index) {
    if (scratchpad.balance - costPerUnitOfEnergy < 0)
      break
    scratchpad.balance -= costPerUnitOfEnergy
    scratchpad.usage++
  }
  return scratchpad
}

const getUsage = (values) => {
  const { monthlySpend, planInfo: { rates, standing_charge = 0 }, vatMultiplier } = values
  const annualCharge = standing_charge * DAYS_IN_YEAR * vatMultiplier
  const annualSpend = monthlySpend * MONTHS_IN_YEAR * PENCE_IN_POUND
  const usage = 0
  const balance = annualSpend - annualCharge
  const scratchpad = rates.reduce(updateUsage, { usage, balance, vatMultiplier })
  return scratchpad.usage
}

const parseData = (options) => {
  const { tokens, context: { plans } }  = options
  if (tokens.length < 4)
    throw new Error('format is "usage <supplier> <plan> <spend>"')
  const [ , supplier, plan, spend ] = tokens
  const findBySupplier = findByMatch.bind(this, 'supplier', supplier)
  const findByPlan = findByMatch.bind(this, 'plan', plan)
  const isValidSpend = isPositiveInteger(spend)
  if (!plans.find(findBySupplier))
    throw new Error(`supplier "${supplier}" not found in existing plans`)
  if (!plans.find(findByPlan))
    throw new Error(`plan "${plan}" not found in existing plans`)
  if (!isValidSpend)
    throw new Error(`spend "${spend}" is not a positive integer amount`)
  return tokens
}

const command = (options) => {
  const { context: { plans, vatMultiplier } }  = options
  const [ , supplier, plan, spend ] = parseData(options)
  const query = { supplier, plan }
  const findByPlan = findByQuery.bind(this, query)
  const monthlySpend = parseInt(spend, 10)
  const planInfo = plans.find(findByPlan)
  if (!planInfo)
    throw new Error(`plan matching query "${JSON.stringify(query)}" not found in existing plans`)
  const values = { vatMultiplier, monthlySpend, planInfo }
  return getUsage(values)
}

export { command }
