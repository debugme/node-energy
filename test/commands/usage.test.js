import { command } from '../../source/commands/usage'
import plans from '../../data/plans'

describe('the usage command function', () => {

  let options = null

  beforeEach(() => {
    options = { tokens: [/* each test to fill in */], context: { plans, 'vatMultiplier': 1.05 } }
  })

  it('should return correct usage for constant plan', () => {
    options.tokens = ['usage', 'bg', 'standing-charge', 120]
    expect(command(options)).toBe(14954)
  })

  it('should return correct usage for variable plan', () => {
    options.tokens = ['usage', 'ovo', 'standard', 1000]
    expect(command(options)).toBe(103855)
  })

  it('should throw an error if wrong number of arguments were passed in', () => {
    options.tokens = ['usage']
    const expectMessage = 'Error: format is "usage <supplier> <plan> <spend>"'
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toEqual(expectMessage)
  })

  it('should throw an error if an unrecognised supplier was passed in', () => {
    options.tokens = ['usage', 'banana', 'variable', 1000]
    const expectMessage = 'Error: supplier "banana" not found in existing plans'
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toEqual(expectMessage)
  })

  it('should throw an error if an unrecognised plan was passed in', () => {
    options.tokens = ['usage', 'eon', 'banana', 1000]
    const expectMessage = 'Error: plan "banana" not found in existing plans'
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toEqual(expectMessage)
  })

  it('should throw an error if invalid monthly spend was passed in ', () => {
    options.tokens = ['usage', 'eon', 'variable', 10.5]
    const expectMessage = 'Error: spend "10.5" is not a positive integer amount'
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toEqual(expectMessage)
  })

  it('should throw an error if invalid supplier and plan were passed in', () => {
    options.tokens = ['usage', 'ovo', 'variable', 10]
    const expectMessage = 'Error: plan matching query "{"supplier":"ovo","plan":"variable"}" not found in existing plans'
    let actualMessage = ''
    try { command(options) }
    catch (error) { actualMessage = error.toString() }
    expect(actualMessage).toEqual(expectMessage)
  })

})
