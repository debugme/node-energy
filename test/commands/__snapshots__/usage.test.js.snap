// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the usage command function should return correct usage for constant plan 1`] = `14954`;

exports[`the usage command function should return correct usage for variable plan 1`] = `103855`;

exports[`the usage command function should throw an error if an unrecognised plan was passed in 1`] = `"Error: plan \\"banana\\" not found in existing plans"`;

exports[`the usage command function should throw an error if an unrecognised supplier was passed in 1`] = `"Error: supplier \\"banana\\" not found in existing plans"`;

exports[`the usage command function should throw an error if invalid monthly spend was passed in  1`] = `"Error: spend \\"10.5\\" is not a positive integer amount"`;

exports[`the usage command function should throw an error if invalid supplier and plan were passed in 1`] = `"Error: plan matching query \\"{\\"supplier\\":\\"ovo\\",\\"plan\\":\\"variable\\"}\\" not found in existing plans"`;

exports[`the usage command function should throw an error if wrong number of arguments were passed in 1`] = `"Error: format is \\"usage <supplier> <plan> <spend>\\""`;
